### PR TITLE
Broker checksum

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request changes the binary format of `messaging.Message` to encode a 4-byte checksum at the beginning of it. This is used when reading data back out to verify that it is not corrupt. Corrupted messages are truncated on recovery so the broker can restart from the previous message.

### CODE REVIEW

File: messaging/broker.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"fmt"
+	"hash/fnv"
```
Comment (otoolep) [inline messaging/broker.go]: Why did you choose this checksum algorithm? I am curious.

Comment (dgnorton) [inline messaging/broker.go]: It's fast

Comment (benbjohnson) [inline messaging/broker.go]: It's fast and it has a good distribution of values: http://programmers.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed

**Comment by toddboom** [comment]
:+1:

**Comment by otoolep** [comment]
Looks good. However I notice there are two places we decode headers and verify that the checksum is OK. Why is that?

File: messaging/broker.go [hunk 2]
```diff
@@ -1232,33 +1233,59 @@ type Message struct {
...
-	copy(b, m.marshalHeader())
-	copy(b[MessageHeaderSize:], m.Data)
+	// Encode message.
+	b := make([]byte, MessageChecksumSize+MessageHeaderSize+len(m.Data))
+	copy(b[MessageChecksumSize:], m.marshalHeader())
+	copy(b[MessageChecksumSize+MessageHeaderSize:], m.Data)
+
+	// Calculate and write the checksum.
+	h := fnv.New32a()
+	h.Write(b[MessageChecksumSize:])
```
Comment (jwilder) [inline messaging/broker.go]: Does this return an error?

File: messaging/broker.go [hunk 3]
```diff
@@ -1232,33 +1233,59 @@ type Message struct {
...
+
+	// Read data.
+	if len(b)-MessageChecksumSize-MessageHeaderSize < len(m.Data) {
+		return fmt.Errorf("message data too short: %d < %d", len(b)-MessageChecksumSize-MessageHeaderSize, len(m.Data))
+	}
+	copy(m.Data, b[MessageChecksumSize+MessageHeaderSize:])
+
+	// Verify checksum.
+	h := fnv.New32a()
+	h.Write(b[MessageChecksumSize:])
```
Comment (jwilder) [inline messaging/broker.go]: Missing error check?

**Comment by benbjohnson** [comment]
@otoolep it's a performance optimization. Sometimes we decode from a `Reader` (e.g. topic files) and sometimes we decode from a `[]byte` (e.g. `LogEntry.Data`). We could consolidate into just the `Reader` and wrap `[]byte` reads in a `bytes.Reader`. I didn't want to change too much in this PR though.

File: messaging/broker.go (hunk 3 above)
Comment (benbjohnson) [inline messaging/broker.go]: The `Write()` functions in `hash/fnv` can't error so I ignore error checks on them:

http://golang.org/src/hash/fnv/fnv.go

**Comment by otoolep** [comment]
OK, thanks @benbjohnson

**Comment by otoolep** [comment]
Oh -- I also meant to ask @benbjohnson -- was the existing message length check insufficient, or you just wanted to be super-safe about corrupted messages?

**Comment by benbjohnson** [comment]
The issue before was a partial write that was then appended to before recovery was done. That issue still needs to get fix but the checksum will ensure the correctness of data that gets read from the broker in the event of any write bug.

**Pull request merged**